### PR TITLE
progress bar for download

### PR DESCRIPTION
**Pull request by lfbdev**
# Summary
add progress bar  when executing download plugins

## Key Points

- [ ] This is a breaking change.
- [ ] Documentation (new or existing) is updated.

## Description

Describe what this PR does and what problems it tries to solve in a few sentences.

### Related Issues

#98

### CODE REVIEW

File: internal/pkg/pluginmanager/pbdownloader.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+	pd.Timeout = time.Second * 60 * 60
+	resp, err := pd.Get(downloadURL)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode == http.StatusOK {
+		log.Printf("[INFO] downloading: [%s]", pluginFilename)
+		log.Print("\n")
```
Comment (daniel-hutao) [inline internal/pkg/pluginmanager/pbdownloader.go:48]: Why not just use `log.Printf("[INFO] downloading: [%s]**\n**", pluginFilename)`

File: internal/pkg/pluginmanager/pbdownloader.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+		defer downFile.Close()
+		// create progress bar when reading response body
+		errSetup := setUpProgressBar(resp, downFile)
+		if errSetup != nil {
+			log.Print(errSetup)
+			return errSetup
+		}
+
+		log.Print("\n")
```
Comment (daniel-hutao) [inline internal/pkg/pluginmanager/pbdownloader.go:63]: Same as above, Print("\n") may be unnecessary

File: internal/pkg/pluginmanager/pbdownloader.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+		log.Printf("[INFO] [%s] download success.", pluginFilename)
+	} else {
+		log.Print("\n")
+		log.Printf("[ERROR] [%s] download fail,%s.", pluginFilename, resp.Status)
```
Comment (daniel-hutao) [inline internal/pkg/pluginmanager/pbdownloader.go:67]: What to do next after the download failed?

File: internal/pkg/pluginmanager/pbdownloader.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	// rename, tmp file to real file
+	err = os.Rename(
```
Comment (daniel-hutao) [inline internal/pkg/pluginmanager/pbdownloader.go:71]: Maybe `if err = os.Rename(); err != nil {} ` is more clear

File: internal/pkg/pluginmanager/pbdownloader.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+		log.Print(err)
+		return err
+	}
+	bar.Finish()
+	return nil
+}
+
+func createPathIfNotExists(path string) error {
+	_, err := os.Stat(path)
+	if err == nil {
```
Comment (daniel-hutao) [inline internal/pkg/pluginmanager/pbdownloader.go:108]: Same here. I'd like to use `if err := func(); err != nil {}`.
There are many places above that have the same problem. Of course, this is not very important, but it is relatively more commonly used.

File: internal/pkg/pluginmanager/pbdownloader.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+		return nil
+	}
+	if os.IsNotExist(err) {
```
Comment (daniel-hutao) [inline internal/pkg/pluginmanager/pbdownloader.go:111]:
```
if !os.IsNotExist(err) {
 	return err
}
if err := os.Mkdir(path, os.ModePerm); err != nil {
 	return err
}
 return nil

**Review by daniel-hutao — CHANGES_REQUESTED** [review]
That's all. Please make sure that the downloader is tested locally.

File: internal/pkg/pluginmanager/pbdownloader.go (hunk 5 above)
Comment (lfbdev) [reply internal/pkg/pluginmanager/pbdownloader.go:108]: done

File: internal/pkg/pluginmanager/pbdownloader.go (hunk 1 above)
Comment (lfbdev) [reply internal/pkg/pluginmanager/pbdownloader.go:48]: done

File: internal/pkg/pluginmanager/pbdownloader.go (hunk 2 above)
Comment (lfbdev) [reply internal/pkg/pluginmanager/pbdownloader.go]: done

File: internal/pkg/pluginmanager/pbdownloader.go (hunk 3 above)
Comment (lfbdev) [reply internal/pkg/pluginmanager/pbdownloader.go:64]: done

File: internal/pkg/pluginmanager/pbdownloader.go (hunk 4 above)
Comment (lfbdev) [reply internal/pkg/pluginmanager/pbdownloader.go]: done

File: internal/pkg/pluginmanager/pbdownloader.go (hunk 6 above)
Comment (lfbdev) [reply internal/pkg/pluginmanager/pbdownloader.go]: done

**Review by IronCore864 — APPROVED** [review]

**Review by daniel-hutao — APPROVED** [review]

**Pull request merged**